ffmpeg-plugin/mcm_common.c

```diff
@@ -4,156 +4,217 @@
+    if (err)
+        return err;
+
+    if (!client) {
+        MeshClientConfig config = { 0 };
+
+        err = mesh_create_client(&client, &config);
+        if (err)
+            client = NULL;
+        else
+            refcnt = 1;
+    } else {
+        refcnt++;
+    }
+
+    pthread_mutex_unlock(&mx);
+
+    *mc = client;
+
+    return err;
+}
+
+/**
+ * Put MCM client, or release the client.
+ * Thread-safe.
+ */
+int mcm_put_client(MeshClient **mc)
+{
+    int err;
 
-    if (type == is_tx)
-        strlcpy(param->remote_addr.port, port, sizeof(param->remote_addr.port));
-    else
-        strlcpy(param->local_addr.port, port, sizeof(param->local_addr.port));
+    err = pthread_mutex_lock(&mx);
+    if (err)
+        return err;
+
+    if (!client) {
+        err = -EINVAL;
+    } else {
+        if (--refcnt == 0)
+            mesh_delete_client(&client);
+
+        *mc = NULL;
+    }
+
+    pthread_mutex_unlock(&mx);
+
+    return err;
+}
+
+/**
+ * Parse MCM connection parameters and fill the structure.
+ */
+int mcm_parse_conn_param(AVFormatContext* avctx, MeshConnection *conn,
+                         int kind, char *ip_addr, char *port,
+                         char *protocol_type, char *payload_type,
+                         char *socket_name, int interface_id)
+{
+    int err;
 
     /* protocol type */
     if (!strcmp(protocol_type, "memif")) {
-        param->protocol = PROTO_MEMIF;
-        snprintf(param->memif_interface.socket_path, sizeof(param->memif_interface.socket_path),
+        MeshConfig_Memif cfg;
+
+        snprintf(cfg.socket_path, sizeof(cfg.socket_path),
             "/run/mcm/mcm_memif_%s.sock", socket_name ? socket_name : "0");
-        param->memif_interface.interface_id = interface_id;
-        param->memif_interface.is_master = (type == is_tx) ? 1 : 0;
-    } else if (!strcmp(protocol_type, "udp")) {
-        param->protocol = PROTO_UDP;
-    } else if (!strcmp(protocol_type, "tcp")) {
-        param->protocol = PROTO_TCP;
-    } else if (!strcmp(protocol_type, "http")) {
-        param->protocol = PROTO_HTTP;
-    } else if (!strcmp(protocol_type, "grpc")) {
-        param->protocol = PROTO_GRPC;
-    } else {
-        param->protocol = PROTO_AUTO;
-    }
+        cfg.interface_id = interface_id;
 
-    /* payload type */
-    if (!strcmp(payload_type, "st20")) {
-        param->payload_type = PAYLOAD_TYPE_ST20_VIDEO;
-    } else if (!strcmp(payload_type, "st22")) {
-        param->payload_type = PAYLOAD_TYPE_ST22_VIDEO;
-        param->payload_codec = PAYLOAD_CODEC_JPEGXS;
-    } else if (!strcmp(payload_type, "st30")) {
-        param->payload_type = PAYLOAD_TYPE_ST30_AUDIO;
-    } else if (!strcmp(payload_type, "st40")) {
-        param->payload_type = PAYLOAD_TYPE_ST40_ANCILLARY;
-    } else if (!strcmp(payload_type, "rtsp")) {
-        param->payload_type = PAYLOAD_TYPE_RTSP_VIDEO;
+        err = mesh_apply_connection_config_memif(conn, &cfg);
+        if (err)
```

Comment:
this could be removed if at the end of function "return err" is done.

Reply:
See my comment above https://github.com/OpenVisualCloud/Media-Communications-Mesh/pull/213#discussion_r1799457946